File: androidApp/app/src/main/kotlin/com/zeitnot/android/setup/SetupPresenter.kt
package com.zeitnot.android.setup

import com.zeitnot.android.domain.Preset
import com.zeitnot.android.domain.PresetCatalogueContract
import com.zeitnot.android.domain.RulesetCategory

data class RulesetModel(
    val id: String,
    val category: RulesetCategory,
    val baseMinutes: Int,
    val incrementSeconds: Int,
    val isSelected: Boolean
)

data class GameConfiguration(
    val category: RulesetCategory,
    val baseMinutes: Int,
    val incrementSeconds: Int
)

class SetupPresenter(catalogue: PresetCatalogueContract) {

    private val presets: List<Preset> = catalogue.presets
    private var selection: Preset = presets.first()

    val rulesetModels: List<RulesetModel>
        get() = presets.map { preset ->
            RulesetModel(
                id = preset.id,
                category = preset.category,
                baseMinutes = preset.baseMinutes,
                incrementSeconds = preset.incrementSeconds,
                isSelected = preset.id == selection.id
            )
        }

    val gameConfiguration: GameConfiguration
        get() = GameConfiguration(
            category = selection.category,
            baseMinutes = selection.baseMinutes,
            incrementSeconds = selection.incrementSeconds
        )

    fun select(id: String) {
        selection = presets.firstOrNull { it.id == id } ?: return
    }

}

File: Shared/Sources/Shared/Setup/SetupPresenter.swift
import Observation

@Observable
public final class SetupPresenter {

    private let router: any SetupRoutingProtocol

    private var selection: PresetRuleset = .bullet1plus0

    public init(router: any SetupRoutingProtocol) {
        self.router = router
    }

    public var rulesetModels: [RulesetModel] {
        PresetRuleset.allCases.map { preset in
            RulesetModel(
                id: preset.id,
                category: preset.category,
                baseMinutes: preset.timeControl.baseMinutes,
                incrementSeconds: preset.timeControl.incrementSeconds,
                isSelected: preset == selection)
        }
    }

    public var startGameModel: StartGameModel {
        StartGameModel(
            category: selection.category,
            baseMinutes: selection.timeControl.baseMinutes,
            incrementSeconds: selection.timeControl.incrementSeconds)
    }

    public var gameConfiguration: GameConfiguration {
        GameConfiguration(timeControl: selection.timeControl, category: selection.category)
    }

    public func select(_ ruleset: PresetRuleset) {
        selection = ruleset
    }

    public func selectRuleset(id: String) {
        guard let ruleset = PresetRuleset(rawValue: id) else { return }

        select(ruleset)
    }

    public func startGame() {
        router.navigateToClock(gameConfiguration: gameConfiguration)
    }

}
